src/gt4py/next/embedded/nd_array_field.py

```diff
@@ -194,6 +211,107 @@ def restrict(self: _BaseNdArrayField, domain) -> _BaseNdArrayField:
 
     __pow__ = _make_binary_array_field_intrinsic_func("pow", "power")
 
+    @overload
+    def __getitem__(self, index: DomainSlice) -> common.Field:
+        """Absolute slicing with dimension names."""
+        ...
+
+    @overload
+    def __getitem__(self, index: tuple[slice | int, ...]) -> common.Field:
+        """Relative slicing with ordered dimension access."""
+        ...
+
+    @overload
+    def __getitem__(
+        self, index: Sequence[common.NamedIndex]
+    ) -> common.Field | core_defs.DType[core_defs.ScalarT]:
+        # Value in case len(i) == len(self.domain)
+        ...
+
+    def __getitem__(
+        self, index: DomainSlice | Sequence[common.NamedIndex] | tuple[int, ...]
+    ) -> common.Field | core_defs.DType[core_defs.ScalarT]:
+        if not isinstance(index, tuple):
+            index = (index,)
+
+        if isinstance(index[0], Domain):
+            index = index[0]
+
+        if is_domain_slice(index):
+            return self._getitem_absolute_slice(index)
+
+        if all(isinstance(idx, (slice, int)) for idx in index):
+            return self._getitem_relative_slice(index)
+
+        raise IndexError(f"Unsupported index type: {index}")
+
+    def _getitem_absolute_slice(self, index: DomainSlice) -> common.Field:
+        slices = _get_slices_from_domain_slice(self.domain, index)
+        new_ranges = []
+        new_dims = []
+        new = self.ndarray[slices]
+
+        for i, dim in enumerate(self.domain.dims):
+            if (pos := _find_index_of_dim(dim, index)) is not None:
+                index_or_range = index[pos][1]
+                if isinstance(index_or_range, UnitRange):
+                    new_ranges.append(index_or_range)
+                    new_dims.append(dim)
+            else:
+                # dimension not mentioned in slice
+                new_ranges.append(self.domain.ranges[i])
+                new_dims.append(dim)
+
+        new_domain = Domain(dims=tuple(new_dims), ranges=tuple(new_ranges))
+
+        return new if new.ndim == 0 else common.field(new, domain=new_domain)
+
+    def _getitem_relative_slice(self, index: tuple[slice | int, ...]) -> common.Field:
+        new = self.ndarray[index]
+        new_dims = []
+        new_ranges = []
+
+        dim_diff = len(new.shape) - len(index)
+
+        if dim_diff > 0:
+            new_index = tuple([*index, Ellipsis])
+        else:
+            new_index = index
+
+        for i, elem in enumerate(new_index):
+            if isinstance(elem, slice):
+                new_dims.append(self.domain.dims[i])
+                new_ranges.append(self._slice_range(self.domain.ranges[i], elem))
+            elif isinstance(elem, int):
+                new_dims.append(self.domain.dims[elem])
+                new_ranges.append(self.domain.ranges[elem])
+            elif isinstance(elem, type(Ellipsis)):
+                new_dims.append(self.domain.dims[i])
+                new_ranges.append(self.domain.ranges[i])
```

Comment:
this is not working in case `len(new.shape) - len(index) < 0`, e.g.` f: Field[[I,J,K]]`, `f[1,:, 2]`.
In my branch I wrote an alternative algorithm, but you could already add test cases for this, and check if it really fails.

Reply:
This should also be fixed now